### DISCUSSION TITLE
Should `current_active_user == None` for logged out users?

### DISCUSSION THREAD

**Discussion by kartzke** [fastapi-users/fastapi-users#1035] [Q&A] [answered]
Hi, I am trying to set up some redirect logic for requests based on the current user's auth status.  Right now, it's a simple login flow:

- Homepage route "/" requires user to be logged in via dependency with `current_active_user`.  I'm using an exception handler to redirect users to "/account"
- Users can log in on the "/account" route
- Once users are logged in, they should be redirected to "/"
- If a logged in user tries to access "/account", they should be redirected to "/".  In the future, I'll show them a "my account" page, but just redirecting to home for now

I have the following routes defined with some redirect logic.  But for logged out users, it seems like `current_active_user` does not equal `None`.  

```
@app.exception_handler(HTTPException)
async def custom_http_exception_handler(request, exc):
    # FastAPI-Users returns a 401 error when user is not logged in, so we catch that one and redirect.
    if exc.status_code == status.HTTP_401_UNAUTHORIZED:
        return RedirectResponse(url='/account')
    # If it's a different status code, use the default error handler.
    return await http_exception_handler(request, exc)


@router.get("/", response_class=HTMLResponse)
async def read_items(request : Request, user: User = Depends(current_active_user)):
    return templates.TemplateResponse("home.html", {"request": request})


@router.get("/account", response_class=HTMLResponse)
async def read_items(request : Request):
    if current_active_user == None:
        return templates.TemplateResponse("users/account.html", {"request": request})
    else:
        return RedirectResponse(url='/')
        
```  

I'm using cookie-based auth with the DB access token.  Without the redirect logic, the auth works perfectly... log in and out are seamless, and the `/users/me` endpoint returns correct user details in the response.

But with this redirect logic, I get an endless 307-redirect loop for logged out users.  If a user is logged in, they always get redirected to the homepage, as expected.  So it seems like the `current_active_user == None` logic doesn't work.  

But as I understand from [the docs](https://fastapi-users.github.io/fastapi-users/10.1/usage/current-user/?h=current_active_user#get-the-current-active-user), it should return `None` for logged out users.

`current_active_user = fastapi_users.current_user()`

I have also tried setting the active flag to true, but it made no difference:  `current_active_user = fastapi_users.current_user(active=True)`

I tried to debug this, but in VS code, but I get the following traceback for the Request -> User variable.  I found some [Starlette docs](https://www.starlette.io/authentication/) regarding this, but I'm not sure how to properly install the middleware with FastAPI-Users.

```
Traceback (most recent call last):\n  
File "/.vscode/extensions/ms-python.python-2022.10.0/pythonFiles/lib/python/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_resolver.py", 
line 192, in _get_py_dictionary\n    attr = getattr(var, name)\n  
File "/envs/py310/lib/python3.10/site-packages/starlette/requests.py", line 160, in user\n    
assert (\nAssertionError: AuthenticationMiddleware must be installed to access request.user\n
```

Any ideas as to why this is looping, or how I can debug this further?

**Comment by frankie567** [comment, marked as answer]
Your issue is related to your previous question, #1034. Typically, you'll need the `optional` parameter to render the home route regardless of the authentication.

```py
optional_current_active_user = fastapi_users.current_user(active=True, optional=True)

@router.get("/", response_class=HTMLResponse)
async def read_items(request : Request, user: Optional[User] = Depends(optional_current_active_user)):
    return templates.TemplateResponse("home.html", {"request": request})
```

**Comment by kartzke** [comment]
Thanks @frankie567 !  That solved the redirect issue, but now when a user is logged in and I check the `/users/me` route, the response contains the HTML of the homepage, not the User data.  My initial thought was this is caused by the mixed response types (HTML and Redirect for the account page), but the route for `user/me` is totally different.  The routes should be isolated, right?

Here is a copy/paste from swagger.

```
curl -X 'GET' \
  'http://127.0.0.1:8000/users/me' \
  -H 'accept: application/json'

Request URL:
http://127.0.0.1:8000/users/me

Server Response
Status Code: 200

Response Body:
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title></title>
    <meta name="description" content="">

# Code below omitted 👇
```

**Reply by frankie567** [reply]
The problem with your automatic redirection is that it'll be catched for every routes, including API like `/users/me`. When making requests in Swagger, the browser automatically follows redirections, that's why you see the response of `/account`.

You probably should avoid having a global exception handler that would catch 401 errors for every routes.

If you really need to make a redirection in some routes, you should probably define a dedicated dependency you can call on specific routes.

**Reply by kartzke** [reply]
Ah, got it, that makes perfect sense.  Ok, thanks for the help with this, much appreciated!
